### DISCUSSION TITLE
RFC: Proposing a Git Development Workflow

### DISCUSSION THREAD

**Discussion by hariDasu** [NJIT-IS332-Fall2023/is322-next-app#3] [Ideas]
**Request for Comment (RFC): Adopting a Git Development Workflow**

**1. Background:**  
As our team and project complexity grow, the need for a well-defined development workflow becomes imperative. This document serves as an RFC for adopting either `Git Flow`, `Trunk Based Development (TBD)`, or a `Hybrid Strategy`.

---

**2. Proposal:**  

**2.1 Git Flow**

*Advantages:*  
- Clear structure for features, releases, and hotfixes.
- Separate branches for features allow for easier collaboration.
- Hotfix branches provide a mechanism to quickly address production issues.

*Disadvantages:*  
- Complexity increases with the number of branches.
- Can be overkill for smaller teams or simpler projects.

*Implementation:*  
1. Develop new features in 'feature/' branches.
2. Merge features into 'develop' when ready for testing.
3. Create 'release/' branches off 'develop' for stabilization.
4. Merge 'release/' into 'master' and tag with a version number.
5. For critical issues in 'master', create 'hotfix/' branches, then merge back into both 'master' and 'develop'.

**2.2 Trunk Based Development (TBD)**

*Advantages:*  
- Simplifies the development process with fewer branches.
- Faster integration and shorter feedback loops.
- Encourages smaller, more frequent releases.

*Disadvantages:*  
- Requires rigorous automated testing to catch regressions.
- Can be challenging for larger features that need longer development times.

*Implementation:*  
1. Developers work in short-lived branches or directly in the trunk.
2. Changes are merged into the trunk frequently, often multiple times a day.
3. Use feature flags to hide incomplete features.
4. Release frequently from the trunk.

**2.3 Hybrid Strategy**

*Advantages:*  
- Combines the structure of Git Flow with the speed and simplicity of TBD.
- Can be tailored to fit the team's specific needs.

*Disadvantages:*  
- Requires careful planning to avoid the pitfalls of both strategies.
- Can become ambiguous if not clearly defined.

*Implementation:*  
1. Use the 'develop' branch as in Git Flow, but merge changes into it more frequently, like in TBD.
2. Use feature branches for longer-lived features, but aim to merge them as soon as possible.
3. Utilize feature flags extensively.
4. Release either from 'develop' or a 'release/' branch, depending on the team's preference.

---

**3. Recommendation:**  
Considering our current team size (3 developers) and the potential for growth in the near future, a flexible approach might be most beneficial. Here's our perspective:

- If our focus was solely on structure and we had a larger team or complex project, `Git Flow` might be more appropriate.
- If we were only prioritizing speed, frequent releases, and had a robust CI/CD pipeline, `TBD` might be best.

Given our unique situation, we suggest adopting the `Hybrid Strategy`. This approach offers the structure needed for growth while maintaining the agility that a smaller team enjoys. As our team evolves, the hybrid approach can be adjusted to best fit our changing needs.

---

**4. Feedback:**  
Please share your thoughts, concerns, and feedback regarding the above proposals. Once we have reached a consensus, we will proceed with implementation and documentation.

---

**Comment by hariDasu** [comment]
this was an old RFC i had written for an earlier archived project, i think its somewhat still valid, though i'd probably opt to change some things since our team here is a bit larger than the original 3 dev mobile team used in this case.  i still think a hybrid flow/tbd strategy is the most scalable way to do remote git with larger teams, so i guess only changes to this would be what is highlighted in this more recent article i (chatgpt) wrote: https://www.dreamstack.us/article/git-flow-strategy/

- `main` as production/stable branch
- `develop` as qa/testing branch
-  release branches cut following our semantic versioning strategy (`release/v1.0.0`) etc
- release branches automated to deploy to `develop` at a fixed time interval 
  -  i.e. weekly, fridays at 11pm or so (open to discussion)

**Reply by benjamin-shuster** [reply]
> release branches automated to deploy to develop at a fixed time interval
    i.e. weekly, fridays at 11pm or so (open to discussion)

I may be misunderstanding, but wouldn't that remove changes that were recently made in `develop`? ie we are making a feature in a feature branch Friday morning, then that night  it gets undone?

**Reply by hariDasu** [reply]
i might have not explained fully in the writing, but how i've used it in the past is after `release/v1.x.x` is cut from develop, feature and bugfix etc branches will be made from the release branch, NOT develop. this way develop is in a weekly "stable" state compared to the release/x branch, which can have merges many times per day possibly
